File: services/message_handler.py
import time
import uuid
import re
import html
import requests
from typing import Dict, List, Optional

class MessageHandler:
    def __init__(self):
        # Rate limiting: track message counts per participant
        # participantId -> { count, resetTime }
        self.rate_limits: Dict[str, dict] = {}
        self.MAX_MESSAGES_PER_WINDOW = 10
        self.RATE_LIMIT_WINDOW = 10 * 1000 # 10 seconds
        self.MAX_MESSAGE_LENGTH = 4000

    def validate_message(self, content: str) -> dict:
        """Validate message content."""
        if not content or not content.strip():
            return {"valid": False, "error": "Message cannot be empty"}

        if len(content) > self.MAX_MESSAGE_LENGTH:
            return {"valid": False, "error": f"Message exceeds maximum length of {self.MAX_MESSAGE_LENGTH} characters"}

        return {"valid": True, "error": None}

    def sanitize_message(self, content: str) -> str:
        """Sanitize message content to prevent XSS."""
        # Escape HTML special characters
        sanitized = html.escape(content)
        
        # Remove any script tags (extra safety)
        sanitized = re.sub(r'<script\b[^<]*(?:(?!</script>)<[^<]*)*</script>', '', sanitized, flags=re.IGNORECASE)
        
        return sanitized

    def check_rate_limit(self, participant_id: str) -> dict:
        """Check rate limit for participant."""
        now = int(time.time() * 1000)
        limit = self.rate_limits.get(participant_id)

        if not limit or now > limit["resetTime"]:
            # Reset or initialize rate limit
            self.rate_limits[participant_id] = {
                "count": 1,
                "resetTime": now + self.RATE_LIMIT_WINDOW
            }
            return {"allowed": True, "error": None}

        if limit["count"] >= self.MAX_MESSAGES_PER_WINDOW:
            return {
                "allowed": False,
                "error": f"Rate limit exceeded. Please wait {max(1, int((limit['resetTime'] - now) / 1000))} seconds."
            }

        # Increment count
        limit["count"] += 1
        return {"allowed": True, "error": None}

    def process_message(self, message: dict, participant_id: str) -> dict:
        """Process and validate message before broadcasting."""
        content = message.get("content", "")
        
        # Validate content
        validation = self.validate_message(content)
        if not validation["valid"]:
            return {"success": False, "message": None, "error": validation["error"]}

        # Check rate limit
        rate_check = self.check_rate_limit(participant_id)
        if not rate_check["allowed"]:
            return {"success": False, "message": None, "error": rate_check["error"]}

        # Sanitize content
        sanitized_content = self.sanitize_message(content)

        # Create processed message
        processed_message = {
            "id": message.get("id") or str(uuid.uuid4()),
            "roomId": message.get("roomId"),
            "senderId": message.get("senderId"),
            "senderName": html.escape(message.get("senderName", "Anonymous")),
            "senderAvatar": message.get("senderAvatar", ""),
            "content": sanitized_content,
            "images": message.get("images", []),
            "timestamp": int(time.time() * 1000),
            "isHelixResponse": message.get("isHelixResponse", False)
        }

        return {"success": True, "message": processed_message, "error": None}

    def cleanup_rate_limits(self) -> int:
        """Clean up old rate limit entries."""
        now = int(time.time() * 1000)
        cleaned_count = 0

        ids_to_delete = []
        for participant_id, limit in self.rate_limits.items():
            if now > limit["resetTime"] + (60 * 1000): # Clean up entries older than 1 minute past reset
                ids_to_delete.append(participant_id)

        for p_id in ids_to_delete:
            del self.rate_limits[p_id]
            cleaned_count += 1

        if cleaned_count > 0:
            print(f"[MessageHandler] Cleaned up {cleaned_count} old rate limit entries")

        return cleaned_count

    def detect_helix_mention(self, content: str) -> bool:
        """Detect if message mentions Helix AI."""
        if not content:
            return False
        
        # Check for "helix" or "@helix" patterns
        return bool(re.search(r'\bhelix\b|^helix|@helix', content, re.IGNORECASE))

    async def forward_to_helix(self, message: dict, room_history: List[dict], participant_count: int, participant_names: List[str] = []) -> str:
        """Forward message to Helix AI backend (non-blocking)."""
        import os
        import asyncio
        
        PORT = os.getenv("PORT", "8000")
        
        def _call_api():
            try:
                # Prepare history
                history = []
                for msg in room_history[-15:]:
                    is_helix = msg.get("isHelixResponse")
                    role = "assistant" if is_helix else "user"
                    # Assistant role doesn't need its name prefixed in history
                    content = msg.get('content') if is_helix else f"{msg.get('senderName')}: {msg.get('content')}"
                    history.append({
                        "role": role,
                        "content": content
                    })

                payload = {
                    "message": message.get("content"),
                    "history": history,
                    "images": message.get("images") if message.get("images") else [],
                    "groupChat": True,
                    "participantNames": participant_names,
                    "participantCount": participant_count
                }

                # Use 127.0.0.1 instead of localhost for more reliable binding resolution
                response = requests.post(f"http://127.0.0.1:{PORT}/api/chat", json=payload, timeout=40)
                response.raise_for_status()
                return response.json().get("reply", "Helix is silent.")
            except Exception as e:
                print(f"[MessageHandler] Threaded API error: {e}")
                return "Helix is temporarily unavailable."

        # Run the blocking requests code in a separate thread
        return await asyncio.to_thread(_call_api)

# Singleton instance
message_handler = MessageHandler()

File: services/MessageHandler.js
const { v4: uuidv4 } = require('uuid');
const validator = require('validator');

/**
 * MessageHandler Service
 * Handles message validation, sanitization, and rate limiting
 */

class MessageHandler {
    constructor() {
        // Rate limiting: track message counts per participant
        this.rateLimits = new Map(); // participantId -> { count, resetTime }
        this.MAX_MESSAGES_PER_WINDOW = 10;
        this.RATE_LIMIT_WINDOW = 10 * 1000; // 10 seconds
        this.MAX_MESSAGE_LENGTH = 4000;
    }

    /**
     * Validate message content
     * @param {string} content - Message content
     * @returns {Object} { valid: boolean, error: string|null }
     */
    validateMessage(content) {
        // Check if empty or whitespace only
        if (!content || validator.isEmpty(content.trim())) {
            return { valid: false, error: 'Message cannot be empty' };
        }

        // Check length
        if (content.length > this.MAX_MESSAGE_LENGTH) {
            return { valid: false, error: `Message exceeds maximum length of ${this.MAX_MESSAGE_LENGTH} characters` };
        }

        return { valid: true, error: null };
    }

    /**
     * Sanitize message content to prevent XSS
     * @param {string} content - Message content
     * @returns {string} Sanitized content
     */
    sanitizeMessage(content) {
        // Escape HTML special characters
        let sanitized = validator.escape(content);
        
        // Remove any script tags (extra safety)
        sanitized = sanitized.replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '');
        
        return sanitized;
    }

    /**
     * Check rate limit for participant
     * @param {string} participantId - Participant identifier
     * @returns {Object} { allowed: boolean, error: string|null }
     */
    checkRateLimit(participantId) {
        const now = Date.now();
        const limit = this.rateLimits.get(participantId);

        if (!limit || now > limit.resetTime) {
            // Reset or initialize rate limit
            this.rateLimits.set(participantId, {
                count: 1,
                resetTime: now + this.RATE_LIMIT_WINDOW
            });
            return { allowed: true, error: null };
        }

        if (limit.count >= this.MAX_MESSAGES_PER_WINDOW) {
            return { 
                allowed: false, 
                error: `Rate limit exceeded. Please wait ${Math.ceil((limit.resetTime - now) / 1000)} seconds.` 
            };
        }

        // Increment count
        limit.count++;
        return { allowed: true, error: null };
    }

    /**
     * Process and validate message before broadcasting
     * @param {Object} message - Raw message object
     * @param {string} participantId - Sender's participant ID
     * @returns {Object} { success: boolean, message: Object|null, error: string|null }
     */
    processMessage(message, participantId) {
        // Validate content
        const validation = this.validateMessage(message.content);
        if (!validation.valid) {
            return { success: false, message: null, error: validation.error };
        }

        // Check rate limit
        const rateCheck = this.checkRateLimit(participantId);
        if (!rateCheck.allowed) {
            return { success: false, message: null, error: rateCheck.error };
        }

        // Sanitize content
        const sanitizedContent = this.sanitizeMessage(message.content);

        // Create processed message
        const processedMessage = {
            id: message.id || uuidv4(),
            roomId: message.roomId,
            senderId: message.senderId,
            senderName: validator.escape(message.senderName || 'Anonymous'),
            senderAvatar: message.senderAvatar || '',
            content: sanitizedContent,
            images: message.images || [],
            timestamp: Date.now(),
            isHelixResponse: message.isHelixResponse || false
        };

        return { success: true, message: processedMessage, error: null };
    }

    /**
     * Clean up old rate limit entries (call periodically)
     */
    cleanupRateLimits() {
        const now = Date.now();
        let cleanedCount = 0;

        for (const [participantId, limit] of this.rateLimits.entries()) {
            if (now > limit.resetTime + (60 * 1000)) { // Clean up entries older than 1 minute past reset
                this.rateLimits.delete(participantId);
                cleanedCount++;
            }
        }

        if (cleanedCount > 0) {
            console.log(`[MessageHandler] Cleaned up ${cleanedCount} old rate limit entries`);
        }

        return cleanedCount;
    }

    /**
     * Detect if message mentions Helix AI
     * @param {string} content - Message content
     * @returns {boolean} True if Helix is mentioned
     */
    detectHelixMention(content) {
        if (!content) return false;
        
        const lowerContent = content.toLowerCase();
        
        // Check for "helix" or "@helix" patterns
        return /\bhelix\b|@helix/i.test(content);
    }

    async forwardToHelix(message, roomHistory, participantCount, participantNames = []) {
        const axios = require('axios');

        try {
            const history = roomHistory.slice(-10).map(msg => ({
                role: msg.isHelixResponse ? 'assistant' : 'user',
                content: `${msg.senderName}: ${msg.content}`
            }));

            const onlineNames = participantNames.filter(Boolean);

            const response = await axios.post('http://localhost:8000/api/chat', {
                message: message.content,
                history: history,
                images: message.images && message.images.length > 0 ? message.images : undefined,
                groupChat: true,
                participantNames: onlineNames,
                participantCount: onlineNames.length,
            });

            return response.data.reply;
        } catch (error) {
            console.error('[MessageHandler] Helix API error:', error.message);
            throw new Error('Helix is temporarily unavailable. Try again in a moment.');
        }
    }
}

// Export singleton instance
const messageHandler = new MessageHandler();

// Schedule periodic cleanup of rate limits (every 5 minutes)
setInterval(() => {
    messageHandler.cleanupRateLimits();
}, 5 * 60 * 1000);

module.exports = messageHandler;
